blockchain/pallets/token_distribution/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            TotalIssuanceOffset::<T>::set(TotalIssuanceOffset::<T>::get() + amount);
+            Ok(())
+        }
+
+        #[pallet::weight((
+            10_000 + T::DbWeight::get().reads_writes(1, 1),
+            DispatchClass::Normal,
+            Pays::No
+        ))]
+        pub fn mint(
+            origin: OriginFor<T>,
+            address: T::AccountId,
+            #[pallet::compact] amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            T::Currency::deposit_creating(&address, amount);
+            TotalIssuanceOffset::<T>::set(TotalIssuanceOffset::<T>::get() + amount);
+
+            Ok(())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn take_from(purpose: u8) -> BalanceOf<T> {
+            PurposeBalances::<T>::take(purpose)
+        }
+
+        pub fn return_to(purpose: u8, amount: BalanceOf<T>) {
+            PurposeBalances::<T>::insert(purpose, amount);
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+    where
+        BalanceOf<T>: num_traits::PrimInt + core::iter::Sum,
+        BlockNumberFor<T>: num_traits::PrimInt,
+    {
+        fn on_finalize(block_number: BlockNumberFor<T>) {
+            let total_weight: u32 = DestinationWeights::<T>::iter_values().sum();
+            if total_weight == 0 {
+                return;
+            }
+
+            let already_issued = T::Currency::total_issuance()
+                + PurposeBalances::<T>::iter_values().sum()
+                - TotalIssuanceOffset::<T>::get();
+
+            // Using Issuance like this makes it _technically_ possible for
+            // consensus to fail if the CPU's floating point calculations are
+            // different.
+            //
+            // If this becomes a problem, we can have this value derived from
+            // an extrinsic that an authoritative account sets. Similar to how
+            // the timestamp pallet works.
+            let issuance = crate::Issuance {
+                total: T::TotalIssuance::get(),
+                half_life: T::IssuanceHalfLife::get(),
+                complete_at: T::IssuanceCompleteAt::get(),
+            };
+            let to_issue = issuance.total_issued_by(block_number) - already_issued;
+            let unit_balance = to_issue / total_weight.into();
+
+            for (dest, weight) in DestinationWeights::<T>::iter() {
+                let to_this = unit_balance * weight.into();
+                match dest {
+                    Destination::Address(a) => {
+                        T::Currency::deposit_creating(&a, to_this);
+                    }
+                    Destination::Purpose(p) => {
+                        let new_balance = PurposeBalances::<T>::get(&p) + to_this;
+                        PurposeBalances::<T>::insert(p, new_balance);
```

Comment:
```suggestion
                        PurposeBalances::<T>::mutate(&p, |a| {
                          *a += to_this;
                        });
```

Can we `mutate` `PurposeBalances` instead?

Reply:
Done.